// probot/lib/octokit/get-probot-octokit-with-defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProbotOctokitWithDefaults = void 0;
const deprecation_1 = require("deprecation");
const get_octokit_throttle_options_1 = require("./get-octokit-throttle-options");
/**
 * Returns an Octokit instance with default settings for authentication. If
 * a `githubToken` is passed explicitly, the Octokit instance will be
 * pre-authenticated with that token when instantiated. Otherwise Octokit's
 * app authentication strategy is used, and `options.auth` options are merged
 * deeply when instantiated.
 *
 * Besides the authentication, the Octokit's baseUrl is set as well when run
 * against a GitHub Enterprise Server with a custom domain.
 */
function getProbotOctokitWithDefaults(options) {
    const authOptions = options.githubToken
        ? {
            token: options.githubToken,
        }
        : {
            cache: options.cache,
            appId: options.appId,
            privateKey: options.privateKey,
        };
    const octokitThrottleOptions = get_octokit_throttle_options_1.getOctokitThrottleOptions({
        log: options.log,
        redisConfig: options.redisConfig,
    });
    if (!options.baseUrl && process.env.GHE_HOST) {
        options.baseUrl = `${process.env.GHE_PROTOCOL || "https"}://${process.env.GHE_HOST}/api/v3`;
        options.log.warn(new deprecation_1.Deprecation(`[probot] "GHE_HOST"/"GHE_PROTOCOL" is deprecated when using with the Probot constructor. Use "new Probot({ baseUrl: 'https://github.acme-inc.com/api/v3' })" instead`));
    }
    const defaultOptions = {
        baseUrl: options.baseUrl,
        auth: authOptions,
    };
    if (options.throttleOptions || octokitThrottleOptions) {
        defaultOptions.throttle = Object.assign({}, options.throttleOptions, octokitThrottleOptions);
    }
    return options.Octokit.defaults((instanceOptions) => {
        const options = Object.assign({}, defaultOptions, instanceOptions, {
            auth: instanceOptions.auth
                ? Object.assign({}, defaultOptions.auth, instanceOptions.auth)
                : defaultOptions.auth,
        });
        if (instanceOptions.throttle) {
            options.throttle = Object.assign({}, defaultOptions.throttle, instanceOptions.throttle);
        }
        return options;
    });
}
exports.getProbotOctokitWithDefaults = getProbotOctokitWithDefaults;
