// probot/lib/create-probot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProbot = void 0;
const deprecation_1 = require("deprecation");
const get_private_key_1 = require("@probot/get-private-key");
const get_log_1 = require("./helpers/get-log");
const probot_1 = require("./probot");
const DEFAULTS = {
    APP_ID: "",
    WEBHOOK_SECRET: "",
    GHE_HOST: "",
    GHE_PROTOCOL: "",
    LOG_FORMAT: "",
    LOG_LEVEL: "warn",
    LOG_LEVEL_IN_STRING: "",
    REDIS_URL: "",
    SENTRY_DSN: "",
};
/**
 * Merges configuration from defaults/environment variables/overrides and returns
 * a Probot instance. Finds private key using [`@probot/get-private-key`](https://github.com/probot/get-private-key).
 *
 * @see https://probot.github.io/docs/configuration/
 * @param defaults default Options, will be overwritten if according environment variable is set
 * @param overrides overwrites defaults and according environment variables
 * @param env defaults to process.env
 */
function createProbot(options = {}) {
    if (isDeprecated(options)) {
        return deprecatedCreateProbot(options);
    }
    const { overrides = {}, defaults = {}, env = process.env } = options;
    const privateKey = get_private_key_1.getPrivateKey({ env });
    const envWithDefaults = { ...DEFAULTS, ...env };
    const envOptions = {
        logLevel: envWithDefaults.LOG_LEVEL,
        appId: Number(envWithDefaults.APP_ID),
        privateKey: (privateKey && privateKey.toString()) || undefined,
        secret: envWithDefaults.WEBHOOK_SECRET,
        redisConfig: envWithDefaults.REDIS_URL,
        baseUrl: envWithDefaults.GHE_HOST
            ? `${envWithDefaults.GHE_PROTOCOL || "https"}://${envWithDefaults.GHE_HOST}/api/v3`
            : "https://api.github.com",
    };
    const probotOptions = {
        ...defaults,
        ...envOptions,
        ...overrides,
    };
    const logOptions = {
        level: probotOptions.logLevel,
        logFormat: envWithDefaults.LOG_FORMAT,
        logLevelInString: envWithDefaults.LOG_LEVEL_IN_STRING === "true",
        sentryDsn: envWithDefaults.SENTRY_DSN,
    };
    const log = get_log_1.getLog(logOptions).child({ name: "server" });
    return new probot_1.Probot({
        log: log.child({ name: "probot" }),
        ...probotOptions,
    });
}
exports.createProbot = createProbot;
function isDeprecated(options) {
    const keys = Object.keys(options);
    return (keys.length > 0 &&
        !keys.includes("overrides") &&
        !keys.includes("defaults") &&
        !keys.includes("env"));
}
function deprecatedCreateProbot(options) {
    options.log =
        options.log ||
            get_log_1.getLog({
                level: process.env.LOG_LEVEL,
                logFormat: process.env.LOG_FORMAT,
                logLevelInString: process.env.LOG_LEVEL_IN_STRING === "true",
                sentryDsn: process.env.SENTRY_DSN,
            });
    const deprecatedKey = Object.keys(options).join(", ");
    options.log.warn(new deprecation_1.Deprecation(`[probot] "createProbot({ ${deprecatedKey} })" is deprecated, use "new Probot(options)" instead.
      
"createProbot(options)" will be repurposed with probot v11 and only accept {defaults, overrides, env} option keys`));
    return new probot_1.Probot(options);
}
