// probot/lib/get-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOptions = void 0;
const deprecation_1 = require("deprecation");
const get_private_key_1 = require("@probot/get-private-key");
const get_log_1 = require("./helpers/get-log");
const probot_1 = require("./probot");
const DEFAULTS = {
    APP_ID: "",
    GHE_HOST: "",
    GHE_PROTOCOL: "",
    HOST: "",
    LOG_FORMAT: "",
    LOG_LEVEL: "warn",
    LOG_LEVEL_IN_STRING: "",
    PORT: "3000",
    REDIS_URL: "",
    SENTRY_DSN: "",
    WEBHOOK_PATH: "",
    WEBHOOK_PROXY_URL: "",
    WEBHOOK_SECRET: "",
};
/**
 * Merges configuration from defaults, environment variables, and overrides.
 * Finds private key using [`@probot/get-private-key`](https://github.com/probot/get-private-key).
 *
 * @see https://probot.github.io/docs/configuration/
 * @param defaults default Options, will be overwritten if according environment variable is set
 * @param overrides overwrites defaults and according environment variables
 * @param env defaults to process.env
 */
function getOptions({ overrides = {}, defaults = {}, env = process.env, } = {}) {
    const privateKey = get_private_key_1.getPrivateKey({ env });
    const envWithDefaults = { ...DEFAULTS, ...env };
    const logOptions = {
        level: envWithDefaults.LOG_LEVEL,
        logFormat: envWithDefaults.LOG_FORMAT,
        logLevelInString: envWithDefaults.LOG_LEVEL_IN_STRING === "true",
        sentryDsn: envWithDefaults.SENTRY_DSN,
    };
    const probotOptions = {
        appId: Number(envWithDefaults.APP_ID),
        privateKey: (privateKey && privateKey.toString()) || undefined,
        secret: envWithDefaults.WEBHOOK_SECRET,
        redisConfig: envWithDefaults.REDIS_URL,
        baseUrl: envWithDefaults.GHE_HOST
            ? `${envWithDefaults.GHE_PROTOCOL || "https"}://${envWithDefaults.GHE_HOST}/api/v3`
            : "https://api.github.com",
    };
    const log = get_log_1.getLog(logOptions).child({ name: "server" });
    const ProbotWithDefaults = probot_1.Probot.defaults({
        ...probotOptions,
        log: log.child({ name: "probot" }),
    });
    const options = {
        ...defaults,
        host: envWithDefaults.HOST,
        port: Number(envWithDefaults.PORT),
        webhookPath: envWithDefaults.WEBHOOK_PATH,
        webhookProxy: envWithDefaults.WEBHOOK_PROXY_URL,
        log,
        Probot: ProbotWithDefaults,
        ...overrides,
    };
    options.log.warn(new deprecation_1.Deprecation(`[probot] "getOptions()" is deprecated, use "{ probot: createProbot() }" instead:

    const { createNodeMiddleware, createProbot } = require("probot");
    const myApp = require("./my-app.js");

    module.exports = createNodeMiddleware(myApp, { probot: createProbot() });`));
    return options;
}
exports.getOptions = getOptions;
